File: src/WebUI/WebShoppingUI/BffCore/Model/Order.cs
namespace WebShoppingUI.Model;

public class Order
{
    public int Id { get; set; }
    public string BuyerId { get; set; }
    public List<OrderItem> Items { get; set; } = new();
    public decimal Total { get; set; }
    public string ShippingAddress { get; set; }
    public OrderStatus Status { get; set; }
    public string? CancelReason { get; set; }
}

public enum OrderStatus
{
    PendingValidation = 1,
    ProcessingPayment = 2,
    PendingSellerConfirmation = 4,
    Shipping = 8,
    Shipped = 16,
    Cancelled = 32,
    Postponed = 64,
}

File: src/WebUI/WebShoppingUI/ClientApp/src/components/Order.js
import React from "react";
import { ApiEndpoints } from "../api/ApiEndpoints";
import { OrderStatusFormatting } from "../constants/OrderStatus.ts";
import { OrderStatus } from "../constants/OrderStatus.ts";

export default function Order({
  orderId,
  items,
  total,
  shippingAddress,
  status,
  cancelReason,
  fnRefreshOrders,
}) {
  const cancellableStatuses = [
    OrderStatus.PendingSellerConfirmation,
    OrderStatus.Postponed,
  ];

  const cancelOrder = async () => {
    var url = ApiEndpoints.orderById(orderId);
    var request = new Request(url, {
      method: "PATCH",
      body: JSON.stringify({
        reason: "waltuh, i'm not doing it waltuh",
      }),
      headers: new Headers({
        "Content-Type": "application/json",
        "X-CSRF": "1",
      }),
    });
    var response = await fetch(request);
    if (response.ok) {
      fnRefreshOrders();
    } else {
      var respBody = await response.json();
      alert(respBody.error);
    }
  };

  return (
    <>
      <tr key={orderId}>
        <td>{orderId}</td>
        <td>
          <div>
            {items.map((i) => (
              <p>{`${i.quantity} x ${i.productName} (${i.productId})`}</p>
            ))}
          </div>
        </td>
        <td>${total}</td>
        <td>{shippingAddress}</td>
        <td>{OrderStatusFormatting[status]}</td>
        <td>{cancelReason ? cancelReason : "-"}</td>
        {cancellableStatuses.includes(status) && (
          <button onClick={cancelOrder}>Cancel</button>
        )}
      </tr>
    </>
  );
}
